operations/secure_random/secure_random.cc

```diff
@@ -0,0 +1,203 @@
+  context->set_output(0, out);
+  return Status::OK();
+}
+
+REGISTER_OP("SecureRandom")
+    .Input("shape: T")
+    .Input("seed: Tseed")
+    .Input("minval: dtype")
+    .Input("maxval: dtype")
+    .Output("output: dtype")
+    .Attr("dtype: {int32, int64} = DT_INT32")
+    .Attr("T: {int32, int64} = DT_INT32")
+    .Attr("Tseed: {int32} = DT_INT32")
+    .SetShapeFn(SecureRandomShape);
+
+// this function allows you to skip ahead in the chacha stream so you don't
+// have to allocate more memory than you need to, used in rejection sampling and
+// will be handy for parallelizing this operation
+void randombytes_buf_deterministic_ic(void * const buf, const size_t size, uint32_t count,
+                              const unsigned char seed[randombytes_SEEDBYTES])
+{
+    static const unsigned char nonce[crypto_stream_chacha20_ietf_NONCEBYTES] = {
+        'L', 'i', 'b', 's', 'o', 'd', 'i', 'u', 'm', 'D', 'R', 'G'
+    };
+
+    unsigned char * u_buf = (unsigned char *)buf;
+
+    memset(u_buf, 0, size);
+
+    crypto_stream_chacha20_ietf_xor_ic(u_buf, u_buf, (unsigned long long) size,
+                                       nonce, count, seed);
+}
+
+
+template <typename T>
+class Generator {
+public:
+  Tensor *output = NULL;
+  const unsigned char * seeds = NULL;
+
+  Generator(Tensor* output, const unsigned char * seeds) : output(output), seeds(seeds) {
+    auto flat = output->flat<T>();
+
+    count_ = output->flat<T>().size();
+    bytes_count_ = count_ * sizeof(T);
+    buf_ = static_cast<T *>(malloc(bytes_count_));
+
+    elements_per_block_ = CHACHABLOCKSIZE / sizeof(T);
+    extra_block_ = static_cast<T *>(malloc(CHACHABLOCKSIZE));
+
+    block_counter_ = bytes_count_ / CHACHABLOCKSIZE + 1;
+
+    // prepare the extra block if any values get rejected in the rejection sampling
+    randombytes_buf_deterministic_ic(extra_block_, CHACHABLOCKSIZE, block_counter_, seeds);
+  }
+
+  ~ Generator() {
+    free(buf_);
+    free(extra_block_);
+  }
+
+  void GenerateData(int minval, int maxval) {    
+    auto flat = output->flat<T>();
+
+    randombytes_buf_deterministic(buf_, bytes_count_, seeds);
+
+    Uniform(minval, maxval);
+
+    std::copy(buf_, buf_ + flat.size(), flat.data());
+  }
+
+private:
+  T *buf_ = NULL;
+  int count_ = 0;
+  int bytes_count_ = 0;
+
+  T * extra_block_ = NULL;
+  uint32_t block_counter_ = 0;
+  int elements_per_block_ = 0;
+  int inner_block_index_ = 0;
+
+  void Uniform(T lo, T hi) {
+    auto range = static_cast<typename std::make_unsigned<T>::type>(hi) -
+                   static_cast<typename std::make_unsigned<T>::type>(lo);
+
+    typename std::make_unsigned<T>::type min = (1U + ~hi) % hi;
+
+    for (int i = 0; i < count_; ++i) {
+      auto unsign = static_cast<typename std::make_unsigned<T>::type>(buf_[i]);
+        while(unsign < min) {
+          // rejection sampling, get the next valid number in the stream
+          buf_[i] = GetNextValidData();
```

Comment:
things got a little crazy here with the rejection sampling. @mortendahl would love for you to give an in depth look into to this and make sure it makes sense! Let me know if you have any ?s.

Reply:
want to add a reference to the method used?

Reply:
rejection sampling was borrowed from here: https://github.com/jedisct1/libsodium/blob/master/src/libsodium/randombytes/randombytes.c#L139